features/mbedtls/targets/TARGET_Cypress/TARGET_PSOC6/aes_alt.c

```diff
@@ -0,0 +1,745 @@
+    AES_VALIDATE_RET( input != NULL );
+    AES_VALIDATE_RET( output != NULL );
+    AES_VALIDATE_RET( mode == MBEDTLS_AES_ENCRYPT ||
+                      mode == MBEDTLS_AES_DECRYPT );
+
+    if( mode == MBEDTLS_AES_ENCRYPT )
+        return( mbedtls_internal_aes_encrypt( ctx, input, output ) );
+    else
+        return( mbedtls_internal_aes_decrypt( ctx, input, output ) );
+}
+
+#if defined(MBEDTLS_CIPHER_MODE_CBC)
+/*
+ * AES-CBC buffer encryption/decryption
+ */
+int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+                    int mode,
+                    size_t length,
+                    unsigned char iv[16],
+                    const unsigned char *input,
+                    unsigned char *output )
+{
+    unsigned char temp[CY_CRYPTO_AES_BLOCK_SIZE];
+
+    int ret = 0;
+    cy_en_crypto_status_t status;
+
+    AES_VALIDATE_RET( ctx != NULL );
+    AES_VALIDATE_RET( mode == MBEDTLS_AES_ENCRYPT ||
+                      mode == MBEDTLS_AES_DECRYPT );
+    AES_VALIDATE_RET( iv != NULL );
+    AES_VALIDATE_RET( input != NULL );
+    AES_VALIDATE_RET( output != NULL );
+
+    if( length % CY_CRYPTO_AES_BLOCK_SIZE )
+        return( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+
+    AES_VALIDATE_RET( ctx->aes_state.buffers != NULL);
+
+    if( mode == MBEDTLS_AES_DECRYPT )
+    {
+        while(( length > 0 ) && (ret == 0))
+        {
+            Cy_Crypto_Core_MemCpy(ctx->obj.base, temp, input, CY_CRYPTO_AES_BLOCK_SIZE);
+
+            status = Cy_Crypto_Core_Aes_Ecb(ctx->obj.base, CY_CRYPTO_DECRYPT, output, input, &ctx->aes_state);
+
+            Cy_Crypto_Core_MemXor(ctx->obj.base, output, output, iv, CY_CRYPTO_AES_BLOCK_SIZE);
+            Cy_Crypto_Core_MemCpy(ctx->obj.base, iv, temp, CY_CRYPTO_AES_BLOCK_SIZE);
+
+            input  += CY_CRYPTO_AES_BLOCK_SIZE;
+            output += CY_CRYPTO_AES_BLOCK_SIZE;
+            length -= CY_CRYPTO_AES_BLOCK_SIZE;
+
+            if (CY_CRYPTO_SUCCESS != status)
+            {
+                ret = MBEDTLS_ERR_AES_HW_ACCEL_FAILED;
+            }
+        }
+    }
+    else
+    {
+        while(( length > 0 ) && (ret == 0))
+        {
+            Cy_Crypto_Core_MemXor(ctx->obj.base, output, input, iv, CY_CRYPTO_AES_BLOCK_SIZE);
+
+            status = Cy_Crypto_Core_Aes_Ecb(ctx->obj.base, CY_CRYPTO_ENCRYPT, output, output, &ctx->aes_state);
+
+            Cy_Crypto_Core_MemCpy(ctx->obj.base, iv, output, CY_CRYPTO_AES_BLOCK_SIZE);
+
+            input  += CY_CRYPTO_AES_BLOCK_SIZE;
+            output += CY_CRYPTO_AES_BLOCK_SIZE;
+            length -= CY_CRYPTO_AES_BLOCK_SIZE;
+
+            if (CY_CRYPTO_SUCCESS != status)
+            {
+                ret = MBEDTLS_ERR_AES_HW_ACCEL_FAILED;
+            }
+        }
+    }
+
+    return( ret );
+}
+
+#endif /* MBEDTLS_CIPHER_MODE_CBC */
+
+#if defined(MBEDTLS_CIPHER_MODE_XTS)
+
+/* Endianess with 64 bits values */
+#ifndef GET_UINT64_LE
+#define GET_UINT64_LE(n,b,i)                            \
```

Comment:
Potential for error if any parameter has a side effect. Also might read better if 'n' was the result of the macro being used as an R-value, eg result = GET_UINT64_LE(buff, offset)

Reply:
This code was captured to support XTS cipher mode from aes.c. Please look into line 1116.